Do not construct solve objects in executioner base class

YaqiWang opened the pull request:
Close #18152.

This is more complicated than what I expected. Even though there are some code duplications, I think this is still the right design. This is needed by Griffin where there are custom executioners do not use `FEProblemSolve`.

While doing this, I moved some code in `Executioner` like grid sequencing and auto scaling into `FEProblemSolve`. I think this is the right thing to do because I do not think wrapping Picard iterations by grid sequencing is the design we want. @lindsayad if this is indeed intentional, I can revert it. Also because of this change, three tests need to be fixed. Two of them can be fixed by adding `accept_on_max_fixed_point_iteration` because now the code is not breaking the grid sequencing loop but the AMR loop causing one output skipped when the maximum number of fixed pointer iterations is hit. The other one I do not have much clue:
```
geomsearch/patch_update_strategy.always-grid-sequencing
```
It exodiffs with a variable `receiver` in the test on `always-grid-sequence_out.e-s002`. I will need @lindsayad 's help here.

This can be reviewed by @GiudGiud.

YaqiWang committed:
do not construct FEProblem solve objects in executioner base class and move the grid sequencing into FEProblem solve object #18152

YaqiWang commented on framework/include/executioners/EigenExecutionerBase.h:
```diff
@@ -131,6 +131,8 @@ class EigenExecutionerBase : public Executioner
   FEProblemBase & _problem;
   MooseEigenSystem & _eigen_sys;
 
+  FEProblemSolve _feproblem_solve;
```
We have to build a dummy solve object for this to-be-outdated executioner otherwise line search, etc. will not be set properly.

GiudGiud commented:
Ok let's talk about this Monday (or earlier if this is time sensitive). I see a lot of duplicated code in the PR and I'm less familiar with Griffin but did you have to add that same code to custom executioners that need fixed point capabilities there too?
Longer term, for these executioners that do not want to have a fixed point solve at all, will they never be used for multiphysics? Or are they purely for transport acceleration?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/764425/) on 4d689ea wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18160/site)

This comment will be updated on new commits.

YaqiWang commented:
There is still `PicardSolve` for multiphysics. They still use MOOSE residual evaluation functions, only do not use `FEProblem::solve` function thus the solve object. A solution to avoid code duplication is to create another layer of inheritance like `FEProblemSolveExecutioner`, which `Steady`, `Eigenvalue` and `Transient` can derive from.

GiudGiud commented:
It looks to me like you moved the fixed point solves to Steady Transient and Eigenvalue. So they're no longer available for the general executioner. 
FEProblemSolve is pretty ubiquitous in moose. What's the file names in Griffin so I can have a look at those non-FEProblemSolve executionners

YaqiWang commented:
Feel free to review MR 261, where the three executioners do not need FEProblemSolve.

GiudGiud commented:
In 261 you are converting these 3 executioners to SolveObject, so they are not going to have a FEProblemSolve. Whatever executioner you are using to solve the griffin problem (I m assuming an Eigenvalue-derived one) will still have a FEProblemSolve after this PR.

YaqiWang commented:
> In 261 you are converting these 3 (4?) executioners to SolveObject, so they are not going to have a FEProblemSolve. One of them is still using a FEProblemSolve btw

Yes, three of them are not using `FEProblemSolve` and another one is doing construction of another solve object (`FEProblemSolve` here) within a solve object (`RichardsonIteration` here).

> 
> What s the main executioner you are planning to use for building these solve objects? I'm assuming the main simulation one. And if you do multiphysics, you will need a SolveObject for the coupling fixed point iterations.

I have a template executioner there that has four instantiations. Currently in that MR, the template executioner is not building `PicardSolve` or any other fixed-point iteration type, but it will for multiphysics. 

> 
> Now if you are trying to alternate CMFD and multiphysics iterations differently than what the fixed point solve allows for, I think you will have to have SolveObjects behave the normal "moose" way where we use the factory, input params, etc. to create/initialize these objects.
> 

The simplest way is to set inner solve of  `PicardSolve` object with the `RichardsonIteration` object. We do not need to have the factory stuff. The template executioner will do things like what we did in `Steady`: build the necessary solve objects and then dial them up. CMFD is another issue. We have talked about how to do multiphysics differently (do not want to talk about in a public domain here).

> Then you could replace the pre-defined moose ones with your custom one in Griffin naturally.
> 

The issue is that that template executioner must be derived from `Executioner`. With our current design in MOOSE, `FEProblemSolve` will be build and its parameters become valid parameters of the template executioner, which are unnecessary.

> It's still not clear to me what you are trying to do. Just reduce memory from not building FEProblemSolve? And you are moving the fixed point solves because they require a FEProblemSolve? They don't, they require an inner solve, which each executioner Steady/Transient/Eigenvalue, set to _feproblem_solve

No, memory is not a concern but user interface. For example `nl_rel_tol` will have no effect in the three executioners but are still a valid parameter. No, fixed point solves or `PicardSolve` is independent on `FEProblemSolve`. I still need to build it in the template executioner (not yet). The template executioner will provide the inner solve object for the `PicardSolve`.

GiudGiud commented:
> I have a template executioner there that has four instantiations. Currently in that MR, the template executioner is not building `PicardSolve` or any other fixed-point iteration type, but it will for multiphysics.

Ok had another look. So it's safe to assume all of these executioners are going to be adding the fixed point solve at some point to tackle multiphysics. I dont see why you would want to get rid of that in the Executioner class. 

> The simplest way is to set inner solve of `PicardSolve` object with the `RichardsonIteration` object. We do not need to have the factory stuff. The template executioner will do things like what we did in `Steady`: build the necessary solve objects and then dial them up. CMFD is another issue. We have talked about how to do multiphysics differently (do not want to talk about in a public domain here).

The factory stuff would allow us to not instantiate a fixed point solve if it's not requested.
You could do that using the factory too. Just add a inner_solve attribute, maybe defaulting to feproblem_solve, to solve objects. 

> The issue is that that template executioner must be derived from `Executioner`. With our current design in MOOSE, `FEProblemSolve` will be build and its parameters become valid parameters of the template executioner, which are unnecessary.

> No, memory is not a concern but user interface. For example `nl_rel_tol` will have no effect in the three executioners but are still a valid parameter. No, fixed point solves or `PicardSolve` is independent on `FEProblemSolve`. I still need to build it in the template executioner (not yet). The template executioner will provide the inner solve object for the `PicardSolve`.

You can disable parameters (suppressParameter) from your derived executioner. The input parameters of the main executioner class are added at the beginning of your derived executioner validParams(), and you can remove them then.
And again, you need the FixedPointSolve object to exist for the templated executioner to provide that inner solve. It can very well be built by the executioner.

YaqiWang commented:
I prefer the non-factory method also because of the user-friendlieness. Normal users do not know much about this objects and neither want to add them through inputs. I do not see absolute reason of using factory to add the solve objects at least for now. I have pretty strong opinion on this. We have went to the suppress parameter path but whenever a new parameter introduced in FEProblemSolve, we will need to add suppressed parameters accordingly and we have been missing lots. The design here is simple: to let concrete executioners add solve objects.

lindsayad commented on framework/include/executioners/Executioner.h:
```diff
@@ -117,19 +117,13 @@ class Executioner : public MooseObject,
-    if (_iteration_method == "picard")
-      return *(dynamic_cast<PicardSolve *>(_fixed_point_solve.get()));
-    else
-      mooseError("Cannot return a PicardSolve if the iteration method is not Picard.");
-  }
+  virtual const PicardSolve * picardSolve() const { return nullptr; }
 
-  FixedPointSolve & fixedPointSolve() { return *_fixed_point_solve; }
+  virtual const FixedPointSolve * fixedPointSolve() const { return nullptr; }
+  virtual FixedPointSolve * fixedPointSolve() { return nullptr; }
```
I really don't like this. Where we were previously assured of always having objects, now we are not. We do not even have any error checking if somehow a derived class doesn't override this and then someone calls `operator->` and gets a seg-fault. Basically I'm not supportive of any design where the default in the base is to error, and in this case it's even worse than erroring, it's seg-faulting.

GiudGiud commented:
The non-factory method does not scale well with the number of options. So it s not that user friendly. It's fine right now, it won't be if we keep adding things there. Which we hopefully will. 
Normal users will keep using actions regardless of whether fixed point solves are an action or not so this does not concern them too much. Some advanced users could benefit from a clean fixed point solve block, where parameters are grouped together and not in the jungle of executioner parameters. Nesting solve objects would be much nicer with the factory syntax. For now, you'd have to rely on hard coding the nesting in either the executioner or an action. 

It's not every day that we add a parameter to FEProblemSolve so I'm not too concerned about a suppression missing. 
The current design also allows you to add solve objects in the executioner. You dont need to modify the framework to do that.

YaqiWang commented:
I decided to keep fixed point iteration solve object in `Executioner`, which is possibly ok because MOOSE is about multiphysics and that object is for multiphysics. This makes the change less dramatic. I still moved the grid sequencing stuff into `FEProblemSolve` thus two changes in the test and one test failure `geomsearch/patch_update_strategy.always-grid-sequencing`.

YaqiWang commented:
>Which we hopefully will.
I doubt we will go that far at least at this point. We possibly will have optimization solve object but we can derive a new executioner for that. I still like to have a fixed set of executioners with which each provides a fixed set of capabilities.

I am very concerned with inactive parameters due to the codes.

YaqiWang commented:
@lindsayad A reminder: you have not commented on this, I do not think wrapping Picard iterations by grid sequencing is the design we want.

lindsayad commented:
> I do not think wrapping Picard iterations by grid sequencing is the design we want.

I agree. Grid sequencing is meant to be a nonlinear preconditioner for Newton

GiudGiud reviewed (changes requested):
NumGridSteps and *solvemethods() need docstrings

I don’t think we need the accept on max fixed point in either of these inputs

Someone needs to understand why the test is now failing before we can consider a regold. Did I miss the explanation in a comment somewhere? I think I saw another thread outside of this PR, not sure where

lindsayad commented:
> Someone needs to understand why the test is now failing before we can consider a regold. Did I miss the explanation in a comment somewhere? I think I saw another thread outside of this PR, not sure where

Yea I agreed that grid sequencing should not be outside of picard iterations ... but grid sequencing should certainly work inside an outer adaptivity loop which is what that test is. That field that is 0 after this PR is measuring something real. It should not be 0. A regold is not acceptable. I will look into what the issue is

moosebuild commented:
Job [Controlled App tests](https://civet.inl.gov/job/765224/) on f2ef0db : invalidated by @YaqiWang

griffin updated

moosebuild commented:
Job [Controlled App tests PETSc alt](https://civet.inl.gov/job/765227/) on f2ef0db : invalidated by @YaqiWang

griffin updated

YaqiWang commented:
The second commit is to show that that is the only test left for fixing. Thanks @lindsayad for looking into it. @GiudGiud check out my message at the top. Basically fixed point iteration hits the maximum number of iterations and returned false. It caused the grid sequencing loop to break (in those two tests, we do not have grid sequencing, but the code has the loop). Then it proceeds to the normal output. With this change here, it caused the AMR loop to break, thus skipped the output resulting in the exodiff. Possibly the test is not properly designed. Do you want to take a look on the tests to see if there is a better way to fix them?

lindsayad committed:
Avoid regold in geomsearch grid sequencing test

The issue is somewhat complex. receiver is an AuxVariable and it's
filled with a boundary restricted aux kernel. Previously aux kernels
executing on timestep end would be executed for each stage of the grid
sequence but now they only execute for the very last grid. Consequently
for the first time step only the nodes on the boundary will have
non-zero values for the aux variable after Yaqi's change. However,
if we change the input to also execute this boundary restricted aux
kernel on nonlinear then we restore the previous behavior with respect
to exodus output:

1. auxkernel is evaluated on the coarse grid
2. refinement occurs and the boundary nodal solution on the coarsest
   grid is projected such that the first layer of interior nodes
   also has non-zero values for the aux variable
3. auxkernel is computed on the finer grid and the boundary nodal
   values are updated
4. solution is output and both the boundary nodes and the first layer
   of interior nodes have non-zero aux variable values

YaqiWang commented:
@lindsayad Thanks for the fix. Looks pretty simple lol. Can you remove the two commits and do a force push? I prefer to keep the changes here clean.

lindsayad commented:
Yep. Just didn't want to force push without permission

GiudGiud commented:
@YaqiWang I ll take a look at those tests in the PR to address  #18201

GiudGiud commented on test/tests/multiapps/picard_multilevel/2level_picard/mutilevel_app.i:
```diff
@@ -69,6 +69,7 @@
   fixed_point_rel_tol = 1E-3
   fixed_point_abs_tol = 1.0e-05
   fixed_point_max_its = 2
+  accept_on_max_fixed_point_iteration = true
```
this one is ok. This test should have allowed for more iterations but with this option the test still meets its purpose

GiudGiud commented on test/tests/multiapps/steffensen/steady_main.i:
```diff
@@ -63,6 +63,7 @@
   fixed_point_algorithm = 'steffensen'
   fixed_point_max_its = 30
   transformed_variables = 'u'
+  accept_on_max_fixed_point_iteration = true
```
this test is also badly designed. It's weird that you are not getting the same issue with the secant method. I ll update this one.

YaqiWang replied:
Sounds great!

YaqiWang commented:
@GiudGiud will you be working on this?

GiudGiud commented:
I am. The tests are failing because of the transform of both the primary and secondary variables, so we ll need to split them. 

Reworked the inner solve a bit in the process, but it broke the secondary variables relaxation so I'm trying to fix it.

GiudGiud reviewed (approved):
@lindsayad any requests?
Temporary workaround for test failures is fine by me, since they ll get huge modifications soon too

lindsayad reviewed (approved):
No requests
